README.md
# Stanford CoreNLP HTTP Server

## About

This software offers the functionality of the [Stanford CoreNLP](http://nlp.stanford.edu/software/corenlp.shtml) as a  HTTP Server. This avoids the time-consuming initialization every time CoreNLP is started. It is very similar to projects like [this Python wrapper](https://github.com/relwell/stanford-corenlp-python).

This is a fork (of a fork) from the original [StanfordCoreNLPXMLServer](https://github.com/nlohmann/StanfordCoreNLPXMLServer) with the followign differences:

- No longer limited to XML - serves JSON/XML depending on the requesting 'Accept' header
- Packaged to run as a server on Debian
	- Listening Interface:Port configured in `/etc/defaults/corenlphttp`
	- Loads configuration from `/etc/corenlphttp/*.properties`

## Example

The server will be listening at <http://localhost:8080>. The text you want to analyze needs to be POSTed as field `text`:

     curl --data 'text=Hello world!' http://localhost:8080

```xml
<?xml version="1.0" encoding="UTF-8"?>
<?xml-stylesheet href="CoreNLP-to-HTML.xsl" type="text/xsl"?>
<root>
  <document>
    <sentences>
      <sentence id="1">
        <tokens>
          <token id="1">
            <word>Hello</word>
            <lemma>hello</lemma>
            <CharacterOffsetBegin>0</CharacterOffsetBegin>
            <CharacterOffsetEnd>5</CharacterOffsetEnd>
            <POS>UH</POS>
            <NER>O</NER>
          </token>
          <token id="2">
            <word>world</word>
            <lemma>world</lemma>
            <CharacterOffsetBegin>6</CharacterOffsetBegin>
            <CharacterOffsetEnd>11</CharacterOffsetEnd>
            <POS>NN</POS>
            <NER>O</NER>
          </token>
          <token id="3">
            <word>!</word>
            <lemma>!</lemma>
            <CharacterOffsetBegin>11</CharacterOffsetBegin>
            <CharacterOffsetEnd>12</CharacterOffsetEnd>
            <POS>.</POS>
            <NER>O</NER>
          </token>
        </tokens>
        <parse>(ROOT (S (VP (NP (INTJ (UH Hello)) (NP (NN world)))) (. !))) </parse>
        <dependencies type="basic-dependencies">
          <dep type="root">
            <governor idx="0">ROOT</governor>
            <dependent idx="2">world</dependent>
          </dep>
          <dep type="discourse">
            <governor idx="2">world</governor>
            <dependent idx="1">Hello</dependent>
          </dep>
        </dependencies>
        <dependencies type="collapsed-dependencies">
          <dep type="root">
            <governor idx="0">ROOT</governor>
            <dependent idx="2">world</dependent>
          </dep>
          <dep type="discourse">
            <governor idx="2">world</governor>
            <dependent idx="1">Hello</dependent>
          </dep>
        </dependencies>
        <dependencies type="collapsed-ccprocessed-dependencies">
          <dep type="root">
            <governor idx="0">ROOT</governor>
            <dependent idx="2">world</dependent>
          </dep>
          <dep type="discourse">
            <governor idx="2">world</governor>
            <dependent idx="1">Hello</dependent>
          </dep>
        </dependencies>
      </sentence>
    </sentences>
  </document>
</root>
```

Note you can also try this [online](http://nlp.stanford.edu:8080/corenlp/process) at Stanford University.
Choose XML or JSON as the ouput format to see something close to how Stanford CoreNLP HTTP Server would respond (with appropriate `Accept` request header)

## Installation

1. Clone the repository:
    
        git clone https://github.com/nlohmann/StanfordCoreNLPXMLServer.git

2. Download and install the third party libraries:
    
        cd StanfordCoreNLPHTTPServer
        ant libs

3. Compile the JAR file:

        ant jar

4. Run the server:

        ant run

5. The server is now waiting on <http://localhost:8080> for HTTP POST requests. Note the initialization can take a few minutes, because several modules and resources of Stanford CoreNLP need to be loaded.

    You can also choose a port:

        ant run -Dport=9000

## Prerequisites

- [Oracle JDK](http://www.oracle.com/technetwork/java/javase/downloads/index.html) or [OpenJDK](http://openjdk.java.net/install/) version 8 or later
- [Apache Ant](http://ant.apache.org)

## Third Party Libraries

The Stanford CoreNLP HTTP Server Server uses the following third party libraries:

- [Stanford CoreNLP](http://nlp.stanford.edu/software/corenlp.shtml), a suite of core NLP tools
- [Simple](http://www.simpleframework.org), a Java based HTTP engine

Stanford CoreNLP is now quired as an external dependency. This is handled automatically by the Debian package.
Simple can be downloaded and set up using the ant target `libs` (see [Installation](#installation)).

## License

- [Stanford CoreNLP](http://nlp.stanford.edu/software/corenlp.shtml) is licensed under the [GNU General Public License (v2 or later)](http://www.gnu.org/licenses/gpl-2.0.html).
- [Simple](http://www.simpleframework.org) is licensed under the [Apache License, Version 2.0](http://www.apache.org/licenses/LICENSE-2.0)

![GNU GPL v3](http://www.gnu.org/graphics/gplv3-127x51.png "GNU GPL v3")

Due to compatibility issues (see [GNU.org](http://www.gnu.org/licenses/license-list.html) and [Apache.org](http://www.apache.org/licenses/GPL-compatibility.html)), the Stanford CoreNLP HTTP Server is licensed under the [**GNU General Public License Version 3**](http://www.gnu.org/licenses/gpl-3.0.html).


debian/changelog
stanford-corenlp-http-server (2.1.0) UNRELEASED; urgency=medium

  * Added RegexNERValidator helper class to quote & test RegexNER maps

 -- Stephen Reay <stephen@koalephant.com>  Tue, 10 Nov 2015 21:14:08 +0700

stanford-corenlp-http-server (2.0.0) unstable; urgency=low

  * Initial release

 -- Stephen Reay <stephen@koalephant.com>  Tue, 03 Nov 2015 21:18:25 +0700


debian/init.d
#!/bin/sh
### BEGIN INIT INFO
# Provides:          stanford-corenlp-http-server
# Required-Start:    $local_fs $network $remote_fs $syslog
# Required-Stop:     $local_fs $network $remote_fs $syslog
# Default-Start:     2 3 4 5
# Default-Stop:      0 1 6
# Short-Description: HTTP Sever for Stanford CoreNLP
# Description:       JSON & XML aware HTTP Server to process text using Stanford CoreNLP
#                    <...>
#                    <...>
### END INIT INFO

# Author: Stephen Reay <stephen@koalephant.com>

# Do NOT "set -e"

# PATH should only include /usr/* if it runs after the mountnfs.sh script
PATH=/sbin:/usr/sbin:/bin:/usr/bin
DESC="HTTP Sever for Stanford CoreNLP"
NAME=stanford-corenlp-http-server
DAEMON=/usr/sbin/stanford-corenlp-http-server
DAEMON_ARGS=""
PIDFILE=/var/run/$NAME.pid
SCRIPTNAME=/etc/init.d/$NAME

# Exit if the package is not installed
[ -x "$DAEMON" ] || exit 0

# Read configuration variable file if it is present
[ -r /etc/default/$NAME ] && . /etc/default/$NAME

# Load the VERBOSE setting and other rcS variables
. /lib/init/vars.sh

# Define LSB log_* functions.
# Depend on lsb-base (>= 3.2-14) to ensure that this file is present
# and status_of_proc is working.
. /lib/lsb/init-functions

#
# Function that starts the daemon/service
#
do_start()
{
	# Return
	#   0 if daemon has been started
	#   1 if daemon was already running
	#   2 if daemon could not be started
	start-stop-daemon --start --quiet --pidfile $PIDFILE --exec $DAEMON --test > /dev/null || return 1
	start-stop-daemon --start --quiet --pidfile $PIDFILE --exec $DAEMON -- $DAEMON_ARGS || return 2
	# The above code will not work for interpreted scripts, use the next
	# six lines below instead (Ref: #643337, start-stop-daemon(8) )
	#start-stop-daemon --start --quiet --pidfile $PIDFILE --startas $DAEMON \
	#	--name $NAME --test > /dev/null \
	#	|| return 1
	#start-stop-daemon --start --quiet --pidfile $PIDFILE --startas $DAEMON \
	#	--name $NAME -- $DAEMON_ARGS \
	#	|| return 2

	# Add code here, if necessary, that waits for the process to be ready
	# to handle requests from services started subsequently which depend
	# on this one.  As a last resort, sleep for some time.
}

#
# Function that stops the daemon/service
#
do_stop()
{
	# Return
	#   0 if daemon has been stopped
	#   1 if daemon was already stopped
	#   2 if daemon could not be stopped
	#   other if a failure occurred
	start-stop-daemon --stop --quiet --retry=TERM/30/KILL/5 --pidfile $PIDFILE --name $NAME
	RETVAL="$?"
	[ "$RETVAL" = 2 ] && return 2
	# Wait for children to finish too if this is a daemon that forks
	# and if the daemon is only ever run from this initscript.
	# If the above conditions are not satisfied then add some other code
	# that waits for the process to drop all resources that could be
	# needed by services started subsequently.  A last resort is to
	# sleep for some time.
	start-stop-daemon --stop --quiet --oknodo --retry=0/30/KILL/5 --exec $DAEMON
	[ "$?" = 2 ] && return 2
	# Many daemons don't delete their pidfiles when they exit.
	rm -f $PIDFILE
	return "$RETVAL"
}

#
# Function that sends a SIGHUP to the daemon/service
#
do_reload() {
	#
	# If the daemon can reload its configuration without
	# restarting (for example, when it is sent a SIGHUP),
	# then implement that here.
	#
	start-stop-daemon --stop --signal 1 --quiet --pidfile $PIDFILE --name $NAME
	return 0
}

case "$1" in
  start)
	[ "$VERBOSE" != no ] && log_daemon_msg "Starting $DESC" "$NAME"
	do_start
	case "$?" in
		0|1) [ "$VERBOSE" != no ] && log_end_msg 0 ;;
		2) [ "$VERBOSE" != no ] && log_end_msg 1 ;;
	esac
	;;
  stop)
	[ "$VERBOSE" != no ] && log_daemon_msg "Stopping $DESC" "$NAME"
	do_stop
	case "$?" in
		0|1) [ "$VERBOSE" != no ] && log_end_msg 0 ;;
		2) [ "$VERBOSE" != no ] && log_end_msg 1 ;;
	esac
	;;
  status)
	status_of_proc "$DAEMON" "$NAME" && exit 0 || exit $?
	;;
  #reload|force-reload)
	#
	# If do_reload() is not implemented then leave this commented out
	# and leave 'force-reload' as an alias for 'restart'.
	#
	#log_daemon_msg "Reloading $DESC" "$NAME"
	#do_reload
	#log_end_msg $?
	#;;
  restart|force-reload)
	#
	# If the "reload" option is implemented then remove the
	# 'force-reload' alias
	#
	log_daemon_msg "Restarting $DESC" "$NAME"
	do_stop
	case "$?" in
	  0|1)
		do_start
		case "$?" in
			0) log_end_msg 0 ;;
			1) log_end_msg 1 ;; # Old process is still running
			*) log_end_msg 1 ;; # Failed to start
		esac
		;;
	  *)
		# Failed to stop
		log_end_msg 1
		;;
	esac
	;;
  *)
	#echo "Usage: $SCRIPTNAME {start|stop|restart|reload|force-reload}" >&2
	echo "Usage: $SCRIPTNAME {start|stop|status|restart|force-reload}" >&2
	exit 3
	;;
esac

:


provision/setup-debian.sh
#!/bin/sh -eu

setup_debian() {

	if [ $(id -u) -ne 0 ]; then
		echo "Must be run as root, exiting"
		return 1
	fi

	setup_apt() {


	echo "Setting up Apt Preferences to disable Suggested package installation"

		cat <<-EOF > /etc/apt/apt.conf.d/99suggests
			APT::Install-Suggests "0";
			Apt::AutoRemove::SuggestsImportant "0";

EOF

		echo "Setting up Apt Preferences to disable Translated package installation"
		cat <<-EOF > /etc/apt/apt.conf.d/99translations
			Acquire::Languages "none";

EOF

		local release="$(lsb_release -c -s)"
		cat <<-EOF > /etc/apt/sources.list
			deb http://ftp.th.debian.org/debian/ ${release} main contrib non-free
			deb http://ftp.th.debian.org/debian/ ${release}-updates main contrib non-free
			deb http://ftp.th.debian.org/debian/ ${release}-backports main contrib non-free
			deb http://security.debian.org ${release}/updates main contrib non-free

EOF

		cat <<-EOT > /etc/apt/preferences.d/${release}-backports
			Package: openjdk-8-jdk openjdk-8-jre openjdk-8-jre-headless openjdk-8-jre-jamvm openjdk-8-jre-zero openjdk-8-source openjdk-8-dbg openjdk-8-demo
			Pin: release a=${release}-backports
			Pin-Priority: 500

EOT

		echo "Updating Apt"
		DEBIAN_FRONTEND=noninteractive aptitude update
		echo ""

		echo "Upgrading base packages"
		DEBIAN_FRONTEND=noninteractive aptitude full-upgrade -y
	}



	install_jdk() {
		echo "Installing Java 8"
		DEBIAN_FRONTEND=noninteractive aptitude -y install aptitude install openjdk-8-jre-headless openjdk-8-jdk ant
	}

	setup_apt "$@"
	install_jdk "$@"
}

setup_debian "$@"


src/StanfordCoreNLPXMLServer.java
/*
    Stanford CoreNLP XML Server
    Copyright (C) 2013 Niels Lohmann

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

import java.io.*;
import java.net.InetSocketAddress;
import java.net.SocketAddress;
import java.util.logging.Level;
import java.util.logging.Logger;

import edu.stanford.nlp.io.*;
import edu.stanford.nlp.ling.*;
import edu.stanford.nlp.pipeline.*;
import edu.stanford.nlp.trees.*;
import edu.stanford.nlp.util.*;

import java.util.concurrent.Callable;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.Future;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.TimeoutException;

import org.simpleframework.http.Request;
import org.simpleframework.http.Response;
import org.simpleframework.http.Query;
import org.simpleframework.http.core.Container;
import org.simpleframework.http.core.ContainerServer;
import org.simpleframework.transport.Server;
import org.simpleframework.transport.connect.Connection;
import org.simpleframework.transport.connect.SocketConnection;

public class StanfordCoreNLPXMLServer implements Container {
    private static StanfordCoreNLP pipeline;
    private static int port = 8080;
    private static final Logger log = Logger.getLogger( StanfordCoreNLPXMLServer.class.getName() );
    private static int total_requests = 0;

    // an interface to the Stanford Core NLP
    public String parse(String s) throws java.io.IOException {
        Annotation annotation = new Annotation(s);
        pipeline.annotate(annotation);
//        try {
//            Thread.sleep(10000);                 //1000 milliseconds is one second.
//        } catch(InterruptedException ex) {
//            Thread.currentThread().interrupt();
//        }
        StringWriter sb = new StringWriter();
        pipeline.xmlPrint(annotation, sb);
        return sb.toString();
    }

    public void handle(Request request, Response response) {
    	int request_number = ++total_requests;
    	final String text = request.getQuery().get("text"); 
        try {
            log.info("Request " + request_number + " from " + request.getClientAddress().getHostName());
            long time = System.currentTimeMillis();
   
            response.setValue("Content-Type", "text/xml");
            response.setValue("Server", "Stanford CoreNLP XML Server/1.0 (Simple 5.1.6)");
            response.setDate("Date", time);
            response.setDate("Last-Modified", time);
   
            // pass "text" POST query to Stanford Core NLP parser 
            PrintStream body = response.getPrintStream();

                       
            // Input Request are handled externally, so here single thread executor is fine
            ExecutorService executor = Executors.newSingleThreadExecutor();
            Callable<Object> task = new Callable<Object>() {
               public Object call() throws IOException {
                  return parse(text);
               }
            };
            Future<Object> future = executor.submit(task);
            Object result = null;
            try {
               result = future.get(60, TimeUnit.SECONDS);
            } catch (TimeoutException ex) {
               // handle the timeout
            	log.log(Level.SEVERE, "TimeoutException", ex);
                log.info("Request " + request_number + ", raised exception for " + text);
            } catch (InterruptedException e) {
               // handle the interrupts
            	log.log(Level.SEVERE, "InterruptedException", e);
                log.info("Request " + request_number + ", raised exception for " + text);
            } catch (ExecutionException e) {
               // handle other exceptions
            	log.log(Level.SEVERE, "ExecutionException", e);
                log.info("Request " + request_number + ", raised exception for " + text);
            } finally {
               future.cancel(true); // may or may not desire this
               executor.shutdownNow();
            }
            body.println(result);
            body.close();
            long time2 = System.currentTimeMillis();
            log.info("Request " + request_number + " done (" + (time2-time) + " ms)");
        } catch(Exception e) {
            log.log(Level.SEVERE, "Exception", e);
            log.info("Request " + request_number + ", raised exception for " + text);
        }
    } 

    public static void main(String args[]) throws Exception {
        // use port if given
        try {
            port = Integer.parseInt(args[0]);
        } catch(Exception e) {
            // silently keep port at 8080
        }

        // initialize the Stanford Core NLP
        pipeline = new StanfordCoreNLP();

        // start the server
        Container container = new StanfordCoreNLPXMLServer();
        Server server = new ContainerServer(container);
        Connection connection = new SocketConnection(server);
        SocketAddress address = new InetSocketAddress(port);
        connection.connect(address);

        log.info("Initialized server at port " + port + ".");
    }
}


src/com/koalephant/util/MediaType.java
/*
	Stanford CoreNLP HTTP Server
	Copyright (C) 2015 Koalephant Co., Ltd

	This program is free software: you can redistribute it and/or modify
	it under the terms of the GNU General Public License as published by
	the Free Software Foundation, either version 3 of the License, or
	(at your option) any later version.

	This program is distributed in the hope that it will be useful,
	but WITHOUT ANY WARRANTY; without even the implied warranty of
	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
	GNU General Public License for more details.

	You should have received a copy of the GNU General Public License
	along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

package com.koalephant.util;

import java.util.regex.Matcher;
import java.util.regex.Pattern;

public enum MediaType {

	TEXT_HTML("text","html"),

	APPLICATION_JSON("application","json"),
	TEXT_JSON("text","json", APPLICATION_JSON),

	TEXT_XML("text","xml"),
	APPLICATION_XML("application","xml", TEXT_XML),
	APPLICATION_XHTML_XML("application", "xhtml+xml"),

	TEXT_JAVASCRIPT("text", "javascript"),
	APPLICATION_JAVASCRIPT("application", "javascript", TEXT_JAVASCRIPT),

	TEXT_CSS("text", "css"),
	TEXT_CSV("text", "csv"),
	TEXT_TSV("text", "tsv"),
	TEXT_PLAIN("text", "plain"),

	IMAGE_GIF("image", "gif"),
	IMAGE_JPG("image", "jpg"),
	IMAGE_JPEG("image", "jpeg", IMAGE_JPG),
	IMAGE_PNG("image", "png");

	/**
	 * Pattern to match type/subtype+suffix
	 */
	private static final Pattern pattern = Pattern.compile("^([^/]+)/([^+;]+)");

	/**
	 * Type
	 */
	private final String type;

	/**
	 * SubType
	 */
	private final String subType;

	/**
	 * The canonical MediaType for this generic type
	 */
	private final MediaType canonical;

	MediaType(String type, String subType) {
		this(type, subType, null);
	}

	MediaType(String type, String subType, MediaType canonical) {
		this.type = type;
		this.subType = subType;
		this.canonical = canonical;
	}

	/**
	 * Get a String representation of this type. e.g. "application/zip"
	 * @return the type/subtype string
	 */
	public String toString() {
		return type + "/" + subType;
	}

	/**
	 * Get the canonical MediaType
	 *
	 * @return the canonical MediaType for this generic type
	 */
	public MediaType getCanonical() {
		return this.canonical != null ? this.canonical : this;
	}

	/**
	 * Check if this is the canonical MediaType for this generic type
	 */
	public boolean isCanonical() {
		return this.canonical == null;
	}

	/**
	 * Check if the given string matches the MediaType
	 * @param typeString the mediatype string to test
	 * @return true/false indicating match
	 */
	public boolean matches(String typeString) throws IllegalArgumentException {
		Matcher parts = pattern.matcher(typeString.toLowerCase());

		if (parts.matches()) {
			String type = parts.group(1);
			String subType = parts.group(2);

			boolean typeMatch = type.equals(this.type) || type.equals("*") || this.type.equals("*");
			boolean subTypeMatch = subType.equals(this.subType) || subType.equals("*") || this.subType.equals("*");

			return typeMatch && subTypeMatch;
		}

		throw new IllegalArgumentException("Cannot parse media type: " + typeString);
	}

	/**
	 * Get a MediaType from a String
	 * @param typeString the mediatype string to lookup
	 * @return the MediaType corresponding to the string provided, or null if none match
	 */
	public static MediaType getFromType(String typeString) {
		return getFromType(typeString, null);
	}

	/**
	 * Get a MediaType from a String
	 * @param typeString the mediatype string to lookup
	 * @param fallback default type to return
	 * @return the MediaType corresponding to the string provided, or the fallback if none match
	 */
	public static MediaType getFromType(String typeString, MediaType fallback) {
		for (MediaType mediaType : MediaType.values()) {
			if(mediaType.matches(typeString)) {
				return mediaType;
			}
		}

		return fallback;
	}
}


src/com/koalephant/nlp/RegexNERValidator.java
/*
	Stanford CoreNLP HTTP Server
	Copyright (C) 2015 Koalephant Co., Ltd

	This program is free software: you can redistribute it and/or modify
	it under the terms of the GNU General Public License as published by
	the Free Software Foundation, either version 3 of the License, or
	(at your option) any later version.

	This program is distributed in the hope that it will be useful,
	but WITHOUT ANY WARRANTY; without even the implied warranty of
	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
	GNU General Public License for more details.

	You should have received a copy of the GNU General Public License
	along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

package com.koalephant.nlp;

import edu.stanford.nlp.pipeline.Annotator;
import edu.stanford.nlp.pipeline.TokensRegexNERAnnotator;
import edu.stanford.nlp.util.StringUtils;
import jdk.nashorn.internal.runtime.regexp.joni.Regex;

import java.io.*;
import java.util.Properties;
import java.util.regex.Pattern;
import java.util.regex.PatternSyntaxException;

/**
 * RegexNER Validator
 */
public class RegexNERValidator {

	/**
	 * Main method
	 *
	 * @param args program arguments, mostly -properties
	 * @throws RuntimeException
	 */
	public static void main(String args[]) throws RuntimeException {

		if (args.length == 0) {
			throw new RuntimeException("No arguments specified");
		}

		Properties properties = StringUtils.argsToProperties(args);

		if (properties.containsKey("inFile")) {
			if (properties.containsKey("outFile")) {
				quoteMappings(properties.getProperty("inFile"), properties.getProperty("outFile"));
			}
			else {
				quoteMappings(properties.getProperty("inFile"));
			}
		}
		else {
			testMappings(properties);
		}
	}

	/**
	 * Test the Mapping by invoking the NERAnnotator
	 *
	 * @param properties the properties object
	 */
	protected static void testMappings(Properties properties) {
		try {
			new TokensRegexNERAnnotator(Annotator.STANFORD_REGEXNER, properties);
		}
		catch (PatternSyntaxException exception) {
			System.err.println(exception.getLocalizedMessage());
			System.exit(1);
		}
	}

	/**
	 * Read an input file, quote the patterns and write to an output file
	 *
	 * @param inFile the file to read in from
	 */
	protected static void quoteMappings(String inFile) {
		quoteMappings(inFile, inFile + ".quoted");
	}

	/**
	 * Read an input file, quote the patterns and write to an output file
	 *
	 * @param inFile the file to read in from
	 * @param outFile the file to write out to
	 */
	protected static void quoteMappings(String inFile, String outFile) {
		BufferedReader bufferedReader = null;
		BufferedWriter bufferedWriter = null;
		try {
			bufferedReader = new BufferedReader(new FileReader(inFile));
			bufferedWriter = new BufferedWriter(new FileWriter(outFile));
			int lineNo = 0;
			String line;
			while ((line = bufferedReader.readLine()) != null) {
				lineNo++;
				String[] tokens = line.split("\t");

				if (tokens.length > 4) {
					System.err.println("Unexpected line longer than 4 tokens: Line " + lineNo);
				}

				String[] patterns = tokens[0].split(" ");
				for (int i = 0; i < patterns.length; i++) {
					patterns[i] = Pattern.quote(patterns[i]);
				}
				tokens[0] = String.join(" ", patterns);

				bufferedWriter.write(String.join("\t", tokens));
				bufferedWriter.newLine();
			}
		}
		catch (FileNotFoundException exception) {
			System.err.println("File not found: " + exception.getLocalizedMessage());
		}
		catch (IOException exception) {
			System.err.println("Error reading file: " + exception.getLocalizedMessage());
		}
		finally {
			if (bufferedReader != null) {
				try {
					bufferedReader.close();
				}
				catch (IOException exception) {
					System.err.println("Error while closing file: " + exception.getLocalizedMessage());
				}
			}

			if (bufferedWriter != null) {
				try {
					bufferedWriter.close();
				}
				catch (IOException exception) {
					System.err.println("Error while closing file: " + exception.getLocalizedMessage());
				}
			}
		}
	}
}


src/com/koalephant/nlp/StanfordCoreNLPHTTPServer.java
/*
	Stanford CoreNLP HTTP Server
	Copyright (C) 2015 Koalephant Co., Ltd

	Original Project:

	Stanford CoreNLP XML Server
	Copyright (C) 2013 Niels Lohmann

	This program is free software: you can redistribute it and/or modify
	it under the terms of the GNU General Public License as published by
	the Free Software Foundation, either version 3 of the License, or
	(at your option) any later version.

	This program is distributed in the hope that it will be useful,
	but WITHOUT ANY WARRANTY; without even the implied warranty of
	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
	GNU General Public License for more details.

	You should have received a copy of the GNU General Public License
	along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

package com.koalephant.nlp;

import com.koalephant.util.MediaType;
import edu.stanford.nlp.ling.CoreAnnotations.DocDateAnnotation;
import edu.stanford.nlp.pipeline.Annotation;
import edu.stanford.nlp.pipeline.StanfordCoreNLP;
import edu.stanford.nlp.util.StringUtils;
import org.joda.time.DateTime;
import org.joda.time.format.DateTimeFormatter;
import org.joda.time.format.ISODateTimeFormat;
import org.simpleframework.http.Request;
import org.simpleframework.http.Response;
import org.simpleframework.http.Status;
import org.simpleframework.http.core.Container;
import org.simpleframework.http.core.ContainerServer;
import org.simpleframework.transport.Server;
import org.simpleframework.transport.connect.Connection;
import org.simpleframework.transport.connect.SocketConnection;

import java.io.IOException;
import java.io.PrintStream;
import java.io.StringWriter;
import java.net.InetSocketAddress;
import java.net.SocketAddress;
import java.util.Properties;
import java.util.concurrent.*;
import java.util.logging.Level;
import java.util.logging.Logger;

/**
 * Stanford CoreNLP HTTP Server
 */
public class StanfordCoreNLPHTTPServer implements Container {

	private static final String serverName = "Stanford CoreNLP HTTP Server/2.0 (Simple 5.1.6)";

	private static final Logger log = Logger.getLogger(StanfordCoreNLPHTTPServer.class.getName());
	private static int total_requests = 0;
	private static String host = "127.0.0.1";
	private static int port = 8080;
	private StanfordCoreNLP pipeline;
	private static MediaType defaultType = MediaType.APPLICATION_JSON;
	private static DateTimeFormatter dateTimeFormatter = ISODateTimeFormat.dateTime();

	/**
	 * Constructor
	 * @param pipeline the StanfordCoreNLP pipeline
	 */
	StanfordCoreNLPHTTPServer(StanfordCoreNLP pipeline) {
		this.pipeline = pipeline;
	}

	/**
	 * Start the Server
	 *
	 * @param args command line arguments
	 * @throws Exception
	 */
	public static void main(String args[]) throws Exception {
		Properties props = new Properties();
		// Load properties from the command line
		if (args.length > 0) {
			log.info("Reading Opts...");
			props = StringUtils.argsToProperties(args);
		}

		// use -host if given
		host = props.getProperty("host", host);
		props.remove("host");

		String portOpt = props.getProperty("port");
		// use -port if given
		if (portOpt != null) {
			try {
				props.remove("port");
				port = Integer.parseInt(portOpt);
			} catch (Exception e) {
				System.err.println("Invalid port specified: " + portOpt);
				System.exit(1);
			}
		}

		String defaultTypeStr = props.getProperty("defaultType");
		if (defaultTypeStr != null) {
			try {
				defaultType = MediaType.getFromType(defaultTypeStr, defaultType);
			}
			catch (IllegalArgumentException e) {
				System.err.println(e.getMessage());
				System.exit(1);
			}
		}

		// start the server
		Container container = new StanfordCoreNLPHTTPServer(new StanfordCoreNLP(props));

		log.info("Attempting to listen on " + host + ":" + port + ".");

		Server server = new ContainerServer(container);
		Connection connection = new SocketConnection(server);
		SocketAddress address = new InetSocketAddress(host, port);
		connection.connect(address);

		log.info("Initialized server at " + host + ":" + port + ".");
	}

	// an interface to the Stanford Core NLP
	public String parse(String s, MediaType mediaType) throws IOException {
		Annotation annotation = new Annotation(s);

		DateTime now = new DateTime();

		annotation.set(DocDateAnnotation.class, now.toString(dateTimeFormatter));
		pipeline.annotate(annotation);
		StringWriter sb = new StringWriter();

		switch (mediaType) {
			case TEXT_XML:
			case APPLICATION_XML:
				pipeline.xmlPrint(annotation, sb);
				break;

			case APPLICATION_JSON:
			case TEXT_JSON:
				pipeline.jsonPrint(annotation, sb);
				break;
		}

		return sb.toString();
	}

	public void handle(Request request, Response response) {
		int request_number = ++total_requests;
		long started = System.currentTimeMillis();

		PrintStream body = null;
		try {
			body = response.getPrintStream();
			if (request.getMethod().equals("HEAD")) {
				response.setStatus(Status.getStatus(200));
				body.close();
				return;
			}
			log.info("Request " + request_number + " from " + request.getClientAddress().getHostName());

			final MediaType mediaType = MediaType.getFromType(request.getValue("Accept"), defaultType);

			final String text = request.getQuery().get("text");

			response.setContentType(mediaType.toString());
			response.setValue("Server", serverName);
			response.setDate("Date", started);
			response.setDate("Last-Modified", started);

			// Input Request are handled externally, so here single thread executor is fine
			ExecutorService executor = Executors.newSingleThreadExecutor();
			Callable<Object> task = () -> parse(text, mediaType);
			Future<Object> future = executor.submit(task);
			Object result = null;
			try {
				result = future.get(60, TimeUnit.SECONDS);
			} catch (TimeoutException ex) {
				// handle the timeout
				log.log(Level.SEVERE, "TimeoutException", ex);
				log.info("Request " + request_number + ", raised exception for " + text);
			} catch (InterruptedException e) {
				// handle the interrupts
				log.log(Level.SEVERE, "InterruptedException", e);
				log.info("Request " + request_number + ", raised exception for " + text);
			} catch (ExecutionException e) {
				// handle other exceptions
				log.log(Level.SEVERE, "ExecutionException", e);
				log.info("Request " + request_number + ", raised exception for " + text);
			} finally {
				future.cancel(true); // may or may not desire this
				executor.shutdownNow();
			}
			long finished = System.currentTimeMillis();
			response.setValue("Processing-Time", String.valueOf(finished - started));
			body.println(result);

			log.info("Request " + request_number + " done (" + (finished - started) + " ms)");
		} catch (Exception e) {
			log.log(Level.SEVERE, "Exception", e);
			log.info("Request " + request_number + ", raised exception");
		}
		finally {
			if (body != null) {
				body.close();
			}
		}
	}
}
